url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/ops/SingularOps.java

```diff
@@ -0,0 +1,375 @@
+                M.set(row,i,M.get(row,bigIndex));
+                M.set(row,bigIndex,tmp);
+            }
+        }
+    }
+
+    /**
+     * <p>
+     * Returns the null-space from the singular value decomposition. The null space is a set of non-zero vectors that
+     * when multiplied by the original matrix return zero.
+     * </p>
+     *
+     * <p>
+     * The null space is found by extracting the columns in V that are associated singular values less than
+     * or equal to the threshold. In some situations a non-compact SVD is required.
+     * </p>
+     *
+     * @param svd A precomputed decomposition.  Not modified.
+     * @param nullSpace Storage for null space.  Will be reshaped as needed.  Modified.
+     * @param tol Threshold for selecting singular values.  Try UtilEjml.EPS.
+     * @return The null space.
+     */
+    public static DenseMatrix64F nullSpace( SingularValueDecomposition<DenseMatrix64F> svd ,
+                                            DenseMatrix64F nullSpace , double tol )
+    {
+        int N = svd.numberOfSingularValues();
+        double s[] = svd.getSingularValues();
+
+        DenseMatrix64F V = svd.getV(null,true);
+
+        if( V.numRows != svd.numCols() ) {
+            throw new IllegalArgumentException("Can't compute the null space using a compact SVD for a matrix of this size.");
+        }
+
+        // first determine the size of the null space
+        int numVectors = svd.numCols()-N;
+
+        for( int i = 0; i < N; i++ ) {
+            if( s[i] <= tol ) {
+                numVectors++;
+            }
+        }
+
+        // declare output data
+        if( nullSpace == null ) {
+            nullSpace = new DenseMatrix64F(numVectors,svd.numCols());
+        } else {
+            nullSpace.reshape(numVectors,svd.numCols());
+        }
+
+        // now extract the vectors
+        int count = 0;
+        for( int i = 0; i < N; i++ ) {
+            if( s[i] <= tol ) {
+                CommonOps.extract(V, i,i+1,0, V.numCols,nullSpace,count++,0);
+            }
+        }
+        for( int i = N; i < svd.numCols(); i++ ) {
+            CommonOps.extract(V, i,i+1,0, V.numCols,nullSpace,count++,0);
+        }
+
+        CommonOps.transpose(nullSpace);
+
+        return nullSpace;
+    }
+
+    /**
+     * <p>
+     * The vector associated will the smallest singular value is returned as the null space
+     * of the decomposed system.  A right null space is returned if 'isRight' is set to true,
+     * and a left null space if false.
+     * </p>
+     *
+     * @param svd A precomputed decomposition.  Not modified.
+     * @param isRight true for right null space and false for left null space.  Right is more commonly used.
+     * @param nullVector Optional storage for a vector for the null space.  Modified.
+     * @return Vector in V associated with smallest singular value..
+     */
+    public static DenseMatrix64F nullVector( SingularValueDecomposition<DenseMatrix64F> svd ,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.